docs/book/tutorials/enabling-vsphere-csi-on-an-existing-cluster.md

```diff
@@ -0,0 +1,133 @@
+export GOVC_PASSWORD=VC_Admin_Passwd
+```
+
+Check the connection to vCenter:
+
+```sh
+$ govc ls
+/datacenter/vm
+/datacenter/network
+/datacenter/host
+/datacenter/datastore
+```
+
+To retrieve all Node VMs, use the following command:
+
+```sh
+$ govc ls /<datacenter-name>/vm
+/datacenter/vm/k8s-node3
+/datacenter/vm/k8s-node4
+/datacenter/vm/k8s-node1
+/datacenter/vm/k8s-node2
+/datacenter/vm/k8s-master
+```
+
+To use govc to enable Disk UUID, use the following command:
+
+```sh
+govc vm.change -vm '/datacenter/vm/k8s-node1' -e="disk.enableUUID=1"
+govc vm.change -vm '/datacenter/vm/k8s-node2' -e="disk.enableUUID=1"
+govc vm.change -vm '/datacenter/vm/k8s-node3' -e="disk.enableUUID=1"
+govc vm.change -vm '/datacenter/vm/k8s-node4' -e="disk.enableUUID=1"
+govc vm.change -vm '/datacenter/vm/k8s-master' -e="disk.enableUUID=1"
+```
+
+Further information on disk.enableUUID can be found in [VMware Knowledgebase Article 52815](https://kb.vmware.com/s/article/52815).
+
+#### Upgrade Virtual Machine Hardware
+
+VM Hardware should be at version 15 or higher.
+
+```bash
+govc vm.upgrade -version=15 -vm '/datacenter/vm/k8s-node1'
+govc vm.upgrade -version=15 -vm '/datacenter/vm/k8s-node2'
+govc vm.upgrade -version=15 -vm '/datacenter/vm/k8s-node3'
+govc vm.upgrade -version=15 -vm '/datacenter/vm/k8s-node4'
+govc vm.upgrade -version=15 -vm '/datacenter/vm/k8s-master'
+```
+
+Check the VM Hardware version after running the above command:
+
+```bash
+$ govc vm.option.info '/datacenter/vm/k8s-node1' | grep HwVersion
+HwVersion:           15
+```
+
+## Kubernetes changes
+
+### Node-level changes
+
+On each K8s node, set the `kubelet`’s `cloud-provider` flag to `external` on all nodes. This flag needs to be set in the service configuration file (usually `/etc/systemd/system/kubelet.service`) but this depends on how you installed Kubernetes or the distribution you are using.
+
+E.g:
+
+```sh
+--cloud-provider=external
+```
+
+Restart the `kubelet` service on each node.
+
+```sh
+systemctl daemon-reload
+systemctl restart kubelet.service
+```
+
+### Kubernetes manifest changes
+
+Set taints on all nodes to allow them to be initialised by the vSphere Cloud Provider Interface, this allows them to have their `providerID` populated, which creates the link between the CSI and the VM in vCenter.
+
+On worker nodes set this taint:
+
+```sh
+kubectl taint nodes <your k8s node name> node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule
+```
+
+On master nodes set this taint:
+
+```sh
+kubectl taint nodes <your k8s master node name> node-role.kubernetes.io/master=:NoSchedule
+```
+
+### Install the vSphere Cloud Provider Interface
+
+Please refer to this guide for details on installing the CPI – <https://github.com/kubernetes/cloud-provider-vsphere/blob/master/docs/book/tutorials/kubernetes-on-vsphere-with-kubeadm.md#install-the-vsphere-cloud-provider-interface>
+
+**Note: Taints needs to be set on the nodes BEFORE the installation of the CPI.**
```

Comment:
I think we actually could apply the taints afterwards as long as we don't start to create volumes prior to force upgrading all nodes

Reply:
Would you like me to remove this note? In my testing the CPI didn't initialise the nodes if they weren't tainted prior to CPI installation - but happy to be told I'm wrong, it would make sense that order is unimportant.

Reply:
I have noticed that as well. This is very true depending on the versions of k8s being run as well. I think it's safer to leave the note in there.